# Notebook A: Data Cleaning and Validation of Publicly Available Clinical Trial Data
## This notebook aims to clean, standardize, and validate the publicly available clinical trial dataset "The systemic responses to oral hyaluronic acid" before statistical analysis.

**Data source:** Zenodo  
**Dataset DOI:** 10.5281/zenodo.20930953

## 1. Library Imports

Let's import the necessary Python libraries for data cleaning and validation.

In [144]:
import pandas as pd
import numpy as np
import re

## 2. Clinical Trial Dataset Import
Let's import the Excel dataset.

The dataset is stored in an Excel file and contains the clinical trial measurements that we will clean and validate in preparation for statistical analysis.

In [145]:
df = pd.read_excel("Data.xlsx")

In [146]:
df.head()

,Treatment,ID,Time,Group,Gender,Phototype,Age,Tape striping on forearm,Tape striping on forearm.1,Tape striping on forearm.2,...,Metabolomics.127,Metabolomics.128,Metabolomics.129,Metabolomics.130,Metabolomics.131,Metabolomics.132,Metabolomics.133,Metabolomics.134,Metabolomics.135,Metabolomics.136
0,Treatment,ID,Time,Group,Gender,Phototype,Age,Alanin,Arginine,Asparagine,...,Ser-Leu-Tyr,Succinic acid,Tetradecanedioic acid,Theophylline,UKOUEEGEHXWHHO-MOFOKWOHSA-N,Undecylenic acid,Uric acid,Uridine,Xanthine,β-Muricholic acid
1,P,101,0,P,1,2,57,294.324357,162.804972,123.093471,...,39668.056,35937221.63,2158096.887,10442046.75,1104819.478,2640774.704,1576279536,9482303.441,7189051.084,69097.41478
2,P,106,0,P,1,2,57,NaN,NaN,NaN,...,34841.32158,28258247.52,9611555.013,22247055.68,1215212.136,2463671.736,1347189090,8784609.248,5360869.54,108788.7305
3,P,111,0,P,1,2,55,420.792795,176.925428,135.474507,...,196565.0648,15589087.99,7125148.718,76887477.94,1321071.949,5008210.167,1419687963,8324165.623,7294962.786,249733.6906
4,P,112,0,P,0,2,34,123.192424,101.040074,78.653531,...,33642.60599,16529940.14,5404199.556,7911626.084,1264210.53,1733370.449,1273509008,6601074.128,4368368.466,319054.0253


## 3. Excel Header Structure
Because the Excel file has a special two-row header, we will use the following code to remove the first header row.

In [147]:
# Import the Excel file without assigning a header
df_raw = pd.read_excel("Data.xlsx", header=None)

# Keep the second row as the actual variable names
column_names = df_raw.iloc[1].astype(str).str.strip()

# Remove both header rows and keep participant data
df = df_raw.iloc[2:].copy()

# Assign the actual variable names
df.columns = column_names

# Reset the index
df.reset_index(drop=True, inplace=True)

In [148]:
df

1,Treatment,ID,Time,Group,Gender,Phototype,Age,Alanin,Arginine,Asparagine,...,Ser-Leu-Tyr,Succinic acid,Tetradecanedioic acid,Theophylline,UKOUEEGEHXWHHO-MOFOKWOHSA-N,Undecylenic acid,Uric acid,Uridine,Xanthine,β-Muricholic acid
0,P,101,0,P,1,2,57,294.324357,162.804972,123.093471,...,39668.056,35937221.63,2158096.887,10442046.75,1104819.478,2640774.704,1576279536,9482303.441,7189051.084,69097.41478
1,P,106,0,P,1,2,57,NaN,NaN,NaN,...,34841.32158,28258247.52,9611555.013,22247055.68,1215212.136,2463671.736,1347189090,8784609.248,5360869.54,108788.7305
2,P,111,0,P,1,2,55,420.792795,176.925428,135.474507,...,196565.0648,15589087.99,7125148.718,76887477.94,1321071.949,5008210.167,1419687963,8324165.623,7294962.786,249733.6906
3,P,112,0,P,0,2,34,123.192424,101.040074,78.653531,...,33642.60599,16529940.14,5404199.556,7911626.084,1264210.53,1733370.449,1273509008,6601074.128,4368368.466,319054.0253
4,P,115,0,P,1,2,55,746.432357,296.462256,135.077298,...,36618.51639,21320749.14,10014310.77,10241510.62,1171966.148,2094599.986,1417415339,10460547.49,5089322.633,59313.78452
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
291,SH60,086,3,SH60,1,2,40,974.120029,664.074258,138.375816,...,35638.6771,22193594.8,9197246.836,32680226.06,1314132.068,2244328.694,1493969075,8578520.07,7155655.166,110894.7083
292,SH60,087,3,SH60,1,2,58,757.310033,403.814963,162.393462,...,44968.0063,17369847.28,12691319.45,19640040.01,1110520.467,2667572.029,1371017431,10474332.88,6225830.879,34839.68278
293,SH60,092,3,SH60,1,2,37,290.541669,118.091931,136.974283,...,32455.35633,25104410.89,10101684.01,14397096.13,1078254.466,1946506.172,1162166310,9071441.04,5456207.725,28487.80866
294,SH60,095,3,SH60,1,2,41,739.514726,599.760443,152.682426,...,34747.50491,27410119.47,12104849.26,24180209.48,41616.50814,1264598.774,1411163519,10043576.74,6790084.64,16382.93961


## 3. Data Cleaning: Preparation and Recoding

The clinical trial dataset was provided as an already cleaned dataset from the source.

No changes were made to the original clinical measurements or participant data.

Only a few minor recoding and formatting modifications were performed to improve readability and facilitate the statistical analysis:

- `P` was recoded to `Placebo` in the Treatment and Group variables.
- Gender coding was converted from `0` = Male and `1` = Female to descriptive labels.
- The Excel header structure was adjusted so that the appropriate variable names could be used in Python.

In [149]:
df["Treatment"] = df["Treatment"].replace("P", "Placebo")
df["Group"] = df["Group"].replace("P", "Placebo")
df["Gender"] = df["Gender"].replace({
    0: "Male",
    1: "Female"
})

Let's check that the recoding worked

In [150]:
print("Treatment groups:")
print(df["Treatment"].value_counts())

print("Groups:")
print(df["Group"].value_counts())

print("\nGender groups:")
print(df["Gender"].value_counts())

Treatment groups:
Treatment
Placebo    100
SH120      100
SH60        96
Name: count, dtype: int64
Groups:
Group
Placebo    100
SH120      100
SH60        96
Name: count, dtype: int64

Gender groups:
Gender
Female    190
Male      106
Name: count, dtype: int64


# The dataset contains two variables, Treatment and Group, which represent the treatment-group information.

After reviewing the variables, Treatment was selected as the main treatment variable because its labels provide a clearer description of the clinical trial groups:

- Placebo
- SH60
- SH120
  
The "Group" column is removed from the analysis dataframe to avoid duplicate information and make the dataset easier to work with.

In [151]:
df = df.drop(columns=["Group"])

In [152]:
df.head()

1,Treatment,ID,Time,Gender,Phototype,Age,Alanin,Arginine,Asparagine,Citruline,...,Ser-Leu-Tyr,Succinic acid,Tetradecanedioic acid,Theophylline,UKOUEEGEHXWHHO-MOFOKWOHSA-N,Undecylenic acid,Uric acid,Uridine,Xanthine,β-Muricholic acid
0,Placebo,101,0,Female,2,57,294.324357,162.804972,123.093471,381.268324,...,39668.056,35937221.63,2158096.887,10442046.75,1104819.478,2640774.704,1576279536,9482303.441,7189051.084,69097.41478
1,Placebo,106,0,Female,2,57,NaN,NaN,NaN,NaN,...,34841.32158,28258247.52,9611555.013,22247055.68,1215212.136,2463671.736,1347189090,8784609.248,5360869.54,108788.7305
2,Placebo,111,0,Female,2,55,420.792795,176.925428,135.474507,704.724983,...,196565.0648,15589087.99,7125148.718,76887477.94,1321071.949,5008210.167,1419687963,8324165.623,7294962.786,249733.6906
3,Placebo,112,0,Male,2,34,123.192424,101.040074,78.653531,163.563132,...,33642.60599,16529940.14,5404199.556,7911626.084,1264210.53,1733370.449,1273509008,6601074.128,4368368.466,319054.0253
4,Placebo,115,0,Female,2,55,746.432357,296.462256,135.077298,823.266482,...,36618.51639,21320749.14,10014310.77,10241510.62,1171966.148,2094599.986,1417415339,10460547.49,5089322.633,59313.78452


### Let's inspect column names.

In [153]:
print(", ".join(df.columns))

Treatment, ID, Time, Gender, Phototype, Age, Alanin, Arginine, Asparagine, Citruline, Fenylalanin, Glutamic acid, glutamin, Glycine, histidin, Isoleucine, kys. asparagova, Leucine, lysin, Methionine, Ornithine, Proline, Pyrrolidone carboxylic acid, Serine, Threonine, Tryptofan, tyrosine, Urea, valin, NMF, Total urocanic acid, trans-Urocanic acid, Hydratation-forehead, Hydratation-cheek, Hydratation-forearm, TEWL-forehead, TEWL-Cheek, Sebum-forehead, Sebum-cheek, Wrinkle, Red-Area-forehead, Red-Area-cheek, Pores-Area-forehead, Pores-Area-cheek, GLOSS-forehead, GLOSS-Cheek, GLOSS-DCS-forehead, GLOSS-DCS-cheek, MEXA Melanin-forehead, MEXA Melanin-cheek, MEXA Erythema-forehead, MEXA Erythema-cheek, ITA-forehead, ITA-cheek, a value-forehead, a value-cheek, b value-forehead, b value-cheek, Elasticity R0-forehead, Elasticity R0-cheek, Elasticity R1-forehead, Elasticity R1-cheek, Elasticity R2-forehead, Elasticity R2-cheek, Elasticity R3-forehead, Elasticity R3-cheek, Elasticity R4 - forehead,


#### The dataset contains a large number of variables that need to be standardized to improve readability and ensure compatibility with SQL and Python conventions. This includes removing unnecessary spaces and special characters and using consistent naming conventions. Let's start with:


- Start column names with a capital letter.
- Replace spaces with _.
- Rename ID → Participant_ID.
- Rename Time → Measurement_Time.
- Rename Kys. Asparagova → Aspartic_Acid.

In [154]:
# Replace spaces with underscores
df.columns = df.columns.str.replace(" ", "_")

# Rename selected columns
df = df.rename(columns={
    "ID": "Participant_ID",
    "Time": "Measurement_Time",
    "kys._asparagova": "Aspartic_Acid"
})

# Start column names with an uppercase letter
df.columns = [col[0].upper() + col[1:] for col in df.columns]

In [155]:
print(", ".join(df.columns))

Treatment, Participant_ID, Measurement_Time, Gender, Phototype, Age, Alanin, Arginine, Asparagine, Citruline, Fenylalanin, Glutamic_acid, Glutamin, Glycine, Histidin, Isoleucine, Aspartic_Acid, Leucine, Lysin, Methionine, Ornithine, Proline, Pyrrolidone_carboxylic_acid, Serine, Threonine, Tryptofan, Tyrosine, Urea, Valin, NMF, Total_urocanic_acid, Trans-Urocanic_acid, Hydratation-forehead, Hydratation-cheek, Hydratation-forearm, TEWL-forehead, TEWL-Cheek, Sebum-forehead, Sebum-cheek, Wrinkle, Red-Area-forehead, Red-Area-cheek, Pores-Area-forehead, Pores-Area-cheek, GLOSS-forehead, GLOSS-Cheek, GLOSS-DCS-forehead, GLOSS-DCS-cheek, MEXA_Melanin-forehead, MEXA_Melanin-cheek, MEXA_Erythema-forehead, MEXA_Erythema-cheek, ITA-forehead, ITA-cheek, A_value-forehead, A_value-cheek, B_value-forehead, B_value-cheek, Elasticity_R0-forehead, Elasticity_R0-cheek, Elasticity_R1-forehead, Elasticity_R1-cheek, Elasticity_R2-forehead, Elasticity_R2-cheek, Elasticity_R3-forehead, Elasticity_R3-cheek, Ela

In [156]:
# Clinical measurement columns to standardize
clinical_columns = [
    "Trans-Urocanic_acid",
    "Hydratation-forehead", "Hydratation-cheek", "Hydratation-forearm",
    "TEWL-forehead", "TEWL-Cheek",
    "Sebum-forehead", "Sebum-cheek",
    "Red-Area-forehead", "Red-Area-cheek",
    "Pores-Area-forehead", "Pores-Area-cheek",
    "GLOSS-forehead", "GLOSS-Cheek",
    "GLOSS-DCS-forehead", "GLOSS-DCS-cheek",
    "MEXA_Melanin-forehead", "MEXA_Melanin-cheek",
    "MEXA_Erythema-forehead", "MEXA_Erythema-cheek",
    "ITA-forehead", "ITA-cheek",
    "A_value-forehead", "A_value-cheek",
    "B_value-forehead", "B_value-cheek",
    "Elasticity_R0-forehead", "Elasticity_R0-cheek",
    "Elasticity_R1-forehead", "Elasticity_R1-cheek",
    "Elasticity_R2-forehead", "Elasticity_R2-cheek",
    "Elasticity_R3-forehead", "Elasticity_R3-cheek",
    "Elasticity_R4_-_forehead", "Elasticity_R4_-_cheek",
    "Elasticity_R5_-_forehead", "Elasticity_R5_-_cheek",
    "Elasticity_R6_-_forehead", "Elasticity_R6_-cheek",
    "Elasticity_R7-forehead", "Elasticity_R7-cheek",
    "Elasticity_R8_-_forehead", "Elasticity_R8_-_cheek",
    "Elasticity_R9-forehead", "Elasticity_R9-cheek",
    "Elasticity_F0-forehead", "Elasticity_F0-cheek",
    "Elasticity_F1-forehead", "Elasticity_F1-cheek",
    "Elasticity_R1_R4-forehead", "Elasticity_R1_R4-cheek",
    "Elasticity_R3_R9-forehead", "Elasticity_R3_R9-cheek",
    "Epidermal_thickness",
]

# Create standardized names only for the selected columns
clinical_rename = {
    col: col.replace("-", "_").replace("Hydratation", "Hydration")
    for col in clinical_columns
}

# Apply the changes
df = df.rename(columns=clinical_rename)

# Start column names with an uppercase letter
df.columns = [col[0].upper() + col[1:] for col in df.columns]

In [157]:
# To verify the selected columns:
print(", ".join(clinical_rename.values()))

Trans_Urocanic_acid, Hydration_forehead, Hydration_cheek, Hydration_forearm, TEWL_forehead, TEWL_Cheek, Sebum_forehead, Sebum_cheek, Red_Area_forehead, Red_Area_cheek, Pores_Area_forehead, Pores_Area_cheek, GLOSS_forehead, GLOSS_Cheek, GLOSS_DCS_forehead, GLOSS_DCS_cheek, MEXA_Melanin_forehead, MEXA_Melanin_cheek, MEXA_Erythema_forehead, MEXA_Erythema_cheek, ITA_forehead, ITA_cheek, A_value_forehead, A_value_cheek, B_value_forehead, B_value_cheek, Elasticity_R0_forehead, Elasticity_R0_cheek, Elasticity_R1_forehead, Elasticity_R1_cheek, Elasticity_R2_forehead, Elasticity_R2_cheek, Elasticity_R3_forehead, Elasticity_R3_cheek, Elasticity_R4___forehead, Elasticity_R4___cheek, Elasticity_R5___forehead, Elasticity_R5___cheek, Elasticity_R6___forehead, Elasticity_R6__cheek, Elasticity_R7_forehead, Elasticity_R7_cheek, Elasticity_R8___forehead, Elasticity_R8___cheek, Elasticity_R9_forehead, Elasticity_R9_cheek, Elasticity_F0_forehead, Elasticity_F0_cheek, Elasticity_F1_forehead, Elasticity_F1_

In [158]:
# Remove repeated underscores from standardized clinical column names
df.columns = [col.replace("__", "_") for col in df.columns]

In [159]:
print(", ".join(df.columns))

Treatment, Participant_ID, Measurement_Time, Gender, Phototype, Age, Alanin, Arginine, Asparagine, Citruline, Fenylalanin, Glutamic_acid, Glutamin, Glycine, Histidin, Isoleucine, Aspartic_Acid, Leucine, Lysin, Methionine, Ornithine, Proline, Pyrrolidone_carboxylic_acid, Serine, Threonine, Tryptofan, Tyrosine, Urea, Valin, NMF, Total_urocanic_acid, Trans_Urocanic_acid, Hydration_forehead, Hydration_cheek, Hydration_forearm, TEWL_forehead, TEWL_Cheek, Sebum_forehead, Sebum_cheek, Wrinkle, Red_Area_forehead, Red_Area_cheek, Pores_Area_forehead, Pores_Area_cheek, GLOSS_forehead, GLOSS_Cheek, GLOSS_DCS_forehead, GLOSS_DCS_cheek, MEXA_Melanin_forehead, MEXA_Melanin_cheek, MEXA_Erythema_forehead, MEXA_Erythema_cheek, ITA_forehead, ITA_cheek, A_value_forehead, A_value_cheek, B_value_forehead, B_value_cheek, Elasticity_R0_forehead, Elasticity_R0_cheek, Elasticity_R1_forehead, Elasticity_R1_cheek, Elasticity_R2_forehead, Elasticity_R2_cheek, Elasticity_R3_forehead, Elasticity_R3_cheek, Elasticit

In [160]:
# Replace special characters in column names with underscores
df.columns = (
    df.columns
    .str.replace(r"[():+\-;]", "_", regex=True)
    .str.replace(r"_+", "_", regex=True)
    .str.strip("_")
)

print(", ".join(df.columns))

Treatment, Participant_ID, Measurement_Time, Gender, Phototype, Age, Alanin, Arginine, Asparagine, Citruline, Fenylalanin, Glutamic_acid, Glutamin, Glycine, Histidin, Isoleucine, Aspartic_Acid, Leucine, Lysin, Methionine, Ornithine, Proline, Pyrrolidone_carboxylic_acid, Serine, Threonine, Tryptofan, Tyrosine, Urea, Valin, NMF, Total_urocanic_acid, Trans_Urocanic_acid, Hydration_forehead, Hydration_cheek, Hydration_forearm, TEWL_forehead, TEWL_Cheek, Sebum_forehead, Sebum_cheek, Wrinkle, Red_Area_forehead, Red_Area_cheek, Pores_Area_forehead, Pores_Area_cheek, GLOSS_forehead, GLOSS_Cheek, GLOSS_DCS_forehead, GLOSS_DCS_cheek, MEXA_Melanin_forehead, MEXA_Melanin_cheek, MEXA_Erythema_forehead, MEXA_Erythema_cheek, ITA_forehead, ITA_cheek, A_value_forehead, A_value_cheek, B_value_forehead, B_value_cheek, Elasticity_R0_forehead, Elasticity_R0_cheek, Elasticity_R1_forehead, Elasticity_R1_cheek, Elasticity_R2_forehead, Elasticity_R2_cheek, Elasticity_R3_forehead, Elasticity_R3_cheek, Elasticit

In [161]:
# Find special characters if still present in column names
special_characters = sorted(set(
    char
    for col in df.columns
    for char in col
    if not char.isalnum() and char != "_"
))

print("Remaining special characters:")
print(special_characters)

Remaining special characters:
["'", ',', '/', '[', ']']


In [162]:
# Remove remaining special characters
# Replace remaining special characters with underscores
df.columns = [
    col.replace("'", "_")
       .replace(",", "_")
       .replace("/", "_")
       .replace("[", "_")
       .replace("]", "_")
    for col in df.columns
]

# Reduce multiple underscores to a single underscore
df.columns = [
    re.sub(r"_+", "_", col).strip("_")
    for col in df.columns
]

# Display all column names for verification
print(", ".join(df.columns))

Treatment, Participant_ID, Measurement_Time, Gender, Phototype, Age, Alanin, Arginine, Asparagine, Citruline, Fenylalanin, Glutamic_acid, Glutamin, Glycine, Histidin, Isoleucine, Aspartic_Acid, Leucine, Lysin, Methionine, Ornithine, Proline, Pyrrolidone_carboxylic_acid, Serine, Threonine, Tryptofan, Tyrosine, Urea, Valin, NMF, Total_urocanic_acid, Trans_Urocanic_acid, Hydration_forehead, Hydration_cheek, Hydration_forearm, TEWL_forehead, TEWL_Cheek, Sebum_forehead, Sebum_cheek, Wrinkle, Red_Area_forehead, Red_Area_cheek, Pores_Area_forehead, Pores_Area_cheek, GLOSS_forehead, GLOSS_Cheek, GLOSS_DCS_forehead, GLOSS_DCS_cheek, MEXA_Melanin_forehead, MEXA_Melanin_cheek, MEXA_Erythema_forehead, MEXA_Erythema_cheek, ITA_forehead, ITA_cheek, A_value_forehead, A_value_cheek, B_value_forehead, B_value_cheek, Elasticity_R0_forehead, Elasticity_R0_cheek, Elasticity_R1_forehead, Elasticity_R1_cheek, Elasticity_R2_forehead, Elasticity_R2_cheek, Elasticity_R3_forehead, Elasticity_R3_cheek, Elasticit

#### Variables Starting With Numbers
Some variable names in the dataset start with a number. Since Python variable names cannot start with a number, let's first identify these column names before deciding how to standardize them.

In [164]:
# Find column names that start with a number 
numeric_start_columns = [ col for col in df.columns if col[0].isdigit() ] 
print("Variables starting with a number:") 
print(numeric_start_columns)

Variables starting with a number:
['12_13_diHOME', '13_HODE', '9_HODE', '5_HETE', '14_15_DiHET', '4E_6E_2_7_dimethyl_8_2R_3R_4S_5S_6R_3_4_5_trihydroxy_6_hydroxymethyl_oxan_2_yl_oxyocta_4_6_dienoic_acid', '4E_7E_dodeca_4_7_dienedioic_acid', '1_16_Hexadecanedioic_acid', '1_3_Dimethyluric_acid', '1_7_Dimethyluric_acid', '10_HDA', '16_Hydroxyhexadecanoic_acid', '1_Methyluric_acid', '1_Methylxanthine', '2_4_Hydroxyphenyl_propanoic_acid', '2_ALPHA_D_MANNOPYRANOSYL_L_TRYPTOPHAN', '2_4_Dinitrophenol', '2_5_Dihydroxybenzenesulfonic_Acid', '2_6_Dihydroxybenzoic_acid', '2_Furoic_acid', '2_Furoylglycine', '2_Hydroxyhippuric_acid', '2_Hydroxyisobutyric_acid', '2_Hydroxyisobutyric_acid', '2_Methyl_4_6_dinitrophenol', '3_1_Pyrazolyl_alanine', '3_Allyl_2_hydroxybenzoic_acid', '3_FUROIC_ACID', '3_Hydroxybutyric_acid', '3_Hydroxydecanoic_acid', '3_Indoxyl_sulphate', '3_Indoxyl_sulphate', '3_Methoxysalicylic_acid', '3_Methoxytyrosine', '3_Methyl_2_oxovaleric_acid_calcium_salt', '3_Methylxanthine', '3_Phe

#### Since these columns represent metabolites, we can add M_ to the beginning of each metabolite variable while preserving the original metabolite name.
For example:
- 12_13_diHOME → M_12_13_diHOME.
- 5_HETE → M_5_HETE.
- 1_Methyluric_acid → M_1_Methyluric_acid.

In [165]:
# Add "M_" to metabolite variables that start with a number 

df.columns = [ "M_" + col if col[0].isdigit() else col for col in df.columns ] 

# Display all column names for verification 

print(", ".join(df.columns))

Treatment, Participant_ID, Measurement_Time, Gender, Phototype, Age, Alanin, Arginine, Asparagine, Citruline, Fenylalanin, Glutamic_acid, Glutamin, Glycine, Histidin, Isoleucine, Aspartic_Acid, Leucine, Lysin, Methionine, Ornithine, Proline, Pyrrolidone_carboxylic_acid, Serine, Threonine, Tryptofan, Tyrosine, Urea, Valin, NMF, Total_urocanic_acid, Trans_Urocanic_acid, Hydration_forehead, Hydration_cheek, Hydration_forearm, TEWL_forehead, TEWL_Cheek, Sebum_forehead, Sebum_cheek, Wrinkle, Red_Area_forehead, Red_Area_cheek, Pores_Area_forehead, Pores_Area_cheek, GLOSS_forehead, GLOSS_Cheek, GLOSS_DCS_forehead, GLOSS_DCS_cheek, MEXA_Melanin_forehead, MEXA_Melanin_cheek, MEXA_Erythema_forehead, MEXA_Erythema_cheek, ITA_forehead, ITA_cheek, A_value_forehead, A_value_cheek, B_value_forehead, B_value_cheek, Elasticity_R0_forehead, Elasticity_R0_cheek, Elasticity_R1_forehead, Elasticity_R1_cheek, Elasticity_R2_forehead, Elasticity_R2_cheek, Elasticity_R3_forehead, Elasticity_R3_cheek, Elasticit

## 4. Data Validation

Before starting the statistical analysis, the prepared dataset was validated to ensure that its structure and variables were appropriate for analysis.

The validation included the following checks:

1. **Participant identification**
   - Confirm the number of unique participants.
   - Check that participant IDs are correctly represented.

2. **Treatment groups**
   - Confirm the presence of the three study groups: Placebo, SH60, and SH120.
   - Check the number of records in each group.

3. **Time**
   - Confirm the available study time points.
   - Verify the distribution of records between baseline and Month 3.

4. **Duplicate records**
   - Check for duplicate participant records.
   - Verify that a participant does not have multiple records for the same time point.

5. **Missing values**
   - Assess missing values across the dataset.
   - Missing values will not be automatically removed or replaced.

6. **Data types**
   - Review whether variables have appropriate data types for statistical analysis.
   - Categorical variables will be treated as categorical variables, while quantitative measurements will be treated as numerical variables where appropriate.

7. **Data consistency**

   
   - Review the dataset for unexpected or inconsistent values and potential outliers.

In [166]:
# Check the number of rows and columns
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 296
Number of columns: 782


In [167]:
# Check participant IDs
print("Total records:", len(df))
print("Unique participants:", df["Participant_ID"].nunique())
print("Missing participant IDs:", df["Participant_ID"].isna().sum())

Total records: 296
Unique participants: 148
Missing participant IDs: 0


In [168]:
# Check measurement time points 
print("Measurement time points:") 
print(df["Measurement_Time"].value_counts(dropna=False).sort_index())

Measurement time points:
Measurement_Time
0    148
3    148
Name: count, dtype: int64


In [169]:
# Check exact duplicate rows
print("Exact duplicate rows:", df.duplicated().sum())

# Check duplicate participant-time combinations
print(
    "Duplicate participant-time records:",
    df.duplicated(
        subset=["Participant_ID", "Measurement_Time"]
    ).sum()
)

Exact duplicate rows: 0
Duplicate participant-time records: 0


Let's check the measurement times for each participant to verify that each participant has measurements at both Time 0 and Time 3.

In [171]:
# Show each participant and their measurement times
print(
    df[["Participant_ID", "Measurement_Time"]]
    .assign(Participant_ID=lambda x: x["Participant_ID"].astype(str))
    .sort_values("Participant_ID")
    .to_string(index=False)
)

Participant_ID Measurement_Time
           001                0
           001                3
           002                0
           002                3
           003                0
           003                3
           004                0
           004                3
           005                0
           005                3
           006                0
           006                3
           007                0
           007                3
           008                0
           008                3
           009                0
           009                3
           010                0
           010                3
           011                0
           011                3
           012                0
           012                3
           013                0
           013                3
           014                0
           014                3
           015                0
           015                3
        

Now we can see that the data contains participants 001–126, then 128–139, then 141–150.
So the missing participant IDs are:127 and 140
which explains the difference: 150 possible participant IDs − 2 missing IDs = 148 actual participants.
And each of the 148 participants has exactly two measurements: Time 0 and Time 3
So there is no problem with the repeated-measures structure based on this output. The IDs 127 and 140 simply do not appear in the dataset.

In [ ]:
# Check for duplicated column names
duplicate_columns = df.columns[df.columns.duplicated()]

print("Number of duplicated column names:", len(duplicate_columns))
print("Duplicated column names:")
print(duplicate_columns.tolist())

In [172]:
# Count every occurrence of each column name
column_counts = {}

for col in df.columns:
    column_counts[col] = column_counts.get(col, 0) + 1

# Keep only names that occur more than once
duplicate_counts = {
    col: count
    for col, count in column_counts.items()
    if count > 1
}

print(duplicate_counts)

{'Glycine': 2, 'LPE_16_0': 2, 'LPE_18_1': 2, 'LPE_22_6': 2, 'TG_50_1': 2, 'M_2_Hydroxyisobutyric_acid': 2, 'M_3_Indoxyl_sulphate': 2, 'CMPF': 2, 'Glycochenodeoxycholic_acid': 3, 'Indole_3_lactic_acid': 2}


In [173]:
# Display only columns with duplicated names, keeping their original order
duplicate_df = df.loc[:, df.columns.duplicated(keep=False)]

duplicate_df

,Glycine,LPE_16_0,LPE_18_1,LPE_22_6,TG_50_1,TG_50_1,M_2_Hydroxyisobutyric_acid,M_2_Hydroxyisobutyric_acid,M_3_Indoxyl_sulphate,M_3_Indoxyl_sulphate,...,CMPF,Glycine,Glycochenodeoxycholic_acid,Glycochenodeoxycholic_acid,Glycochenodeoxycholic_acid,Indole_3_lactic_acid,Indole_3_lactic_acid,LPE_16_0,LPE_18_1,LPE_22_6
0,487.924908,0.000153,0.000161,0.000132,0.015397,0.000623,11992157.7,226233931.3,12849489.38,272670736.3,...,493064.2822,4604834.453,1716821.38,6173172.559,2132732.318,20734778.78,434817.8721,2791114.628,5123934.489,693742.0321
1,NaN,0.000162,0.000141,0.000123,0.008857,0.000445,16002983.15,175352282,16725141.21,329799588,...,739203.8487,8276117.379,466223.3659,1234978.996,1580375.992,18521822.66,3151552.538,4478026.474,6112722.698,860610.9159
2,745.483216,0.000204,0.000157,0.000143,0.039254,0.001302,20265786.3,360557567.6,10231158.4,245090466.5,...,304406.0782,3486505.078,1015728.261,5067980,3150277.732,29099202.54,2747768.995,3995242.211,4772666.402,803603.0012
3,193.949239,0.000446,0.000361,0.000373,0.015501,0.000938,17293806.92,172911056.8,8507377.088,204723863.2,...,520870.9911,4725040.5,56742766.12,20701762.25,1573194.469,21590088.39,239363.7576,5974337.903,8029934.221,728035.9214
4,882.377623,0.000194,0.000174,0.000115,0.012404,0.000516,15290513.64,167628988.9,8278041.119,202150666.7,...,284478.3373,4289720.039,1208344.369,4852248.234,5481488.155,16262188.65,3197246.479,3230090.748,5022128.99,545964.0845
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
291,1076.643226,0.000168,0.000148,0.000112,0.016078,0.000829,17818888.34,227421258.8,11825523.97,262835842.9,...,1581463.415,4350400.631,731973.4538,11646921.66,1656136.311,30691471.66,2710774.987,4198715.737,5806323.196,800380.7569
292,983.387411,0.000225,0.000205,0.000073,0.011697,0.000505,15428606.66,183576077,14082236.71,343919797.1,...,572111.1277,3321495.93,11555591.8,19014910.87,5461246.397,23274352.22,2183080.868,6484089.787,10435720.85,744325.0024
293,550.420421,0.000143,0.000213,0.00007,0.007895,0.000483,11953088.42,98785748.87,16498211.69,459041359.2,...,1132741.214,3854500.574,4402977.44,10614428.68,1852478.612,22414289.48,5654812.001,4299640.742,9253863.653,676471.3947
294,1097.427736,0.000238,0.000217,0.000188,0.007434,0.000474,13115456.59,157570887.4,14392613.39,334868096.7,...,866777.8594,4897819.718,2344360.33,7633358.004,12185047.41,26785256.85,3083442.714,4272363.182,5503512.004,739262.651


Some column names appear more than once in the dataset, while the corresponding columns contain different values. The study reports untargeted metabolomic and lipidomic analyses, but does not explain the repeated column names found in the public dataset. These columns are therefore retained for further analysis.

# Missing Values

Let's check for missing values in each variable to identify incomplete measurements in the dataset.

In [174]:
# Check missing values in each column
missing_values = df.isna().sum()

# Display only columns with missing values
missing_values = missing_values[missing_values > 0]

print("Columns with missing values:")
print(missing_values)

Columns with missing values:
Alanin               8
Arginine             8
Asparagine           8
Citruline            8
Fenylalanin          8
                    ..
Undecylenic_acid     4
Uric_acid            4
Uridine              4
Xanthine             4
Β_Muricholic_acid    4
Length: 776, dtype: int64


In [175]:
# Calculate the number and percentage of missing values in each column
missing_summary = pd.DataFrame({
    "Missing_Count": df.isna().sum(),
    "Missing_Percent": df.isna().mean() * 100
})

# Keep only columns with missing values
missing_summary = missing_summary[missing_summary["Missing_Count"] > 0]

# Display the summary
missing_summary

,Missing_Count,Missing_Percent
Alanin,8,2.702703
Arginine,8,2.702703
Asparagine,8,2.702703
Citruline,8,2.702703
Fenylalanin,8,2.702703
...,...,...
Undecylenic_acid,4,1.351351
Uric_acid,4,1.351351
Uridine,4,1.351351
Xanthine,4,1.351351


There are 776 variables with at least one missing value.

In [177]:
# Calculate the percentage of missing values in each column
missing_percent = df.isna().mean() * 100

# Identify columns with at least 50% missing values
high_missing_columns = missing_percent[missing_percent >= 50].sort_values(
    ascending=False
)

print("Number of columns with at least 50% missing values:", len(high_missing_columns))
print("\nColumns with at least 50% missing values:")
print(high_missing_columns)

Number of columns with at least 50% missing values: 4

Columns with at least 50% missing values:
C_peptide    66.891892
IL6          66.891892
Leptin       66.891892
MCP1         66.891892
dtype: float64


In [178]:
# Count non-missing measurements for each treatment group and time point
available_entries = (
    df.groupby(["Treatment", "Measurement_Time"])[high_missing_columns.index]
      .count()
)

available_entries

C_peptide  IL6  Leptin  MCP1
Treatment Measurement_Time                              
Placebo   0                         0    0       0     0
          3                         0    0       0     0
SH120     0                        49   49      49    49
          3                        49   49      49    49
SH60      0                         0    0       0     0
          3                         0    0       0     0

These four variables contain measurements only for the SH120 treatment group (49 observations at both Measurement Time 0 and 3), while no measurements are available for the Placebo or SH60 groups. Therefore, these variables cannot be used for comparisons across all three treatment groups and will be excluded from the statistical analysis.

In [179]:
# Check for variables with data in only one treatment group
# Columns already excluded
excluded_columns = ["C_peptide", "IL6", "Leptin", "MCP1"]

found = 0

for i, col in enumerate(df.columns):

    if col in ["Participant_ID", "Treatment", "Measurement_Time"]:
        continue

    if col in excluded_columns:
        continue

    values = df.iloc[:, i]

    groups_with_data = df.assign(
        _value=values
    ).groupby("Treatment")["_value"].count()

    if (groups_with_data > 0).sum() == 1:
        print(col)
        print(groups_with_data)
        print()
        found += 1

print("Additional columns found:", found)

Additional columns found: 0


In [180]:
# Variables already excluded from statistical analysis
excluded_columns = ["C_peptide", "IL6", "Leptin", "MCP1"]

# Get measurement columns
measurement_columns = [
    col for col in df.columns
    if col not in ["Participant_ID", "Treatment", "Measurement_Time"]
    and col not in excluded_columns
]

print("Number of measurement variables:", len(measurement_columns))

Number of measurement variables: 775


In [181]:
# Variables already excluded from statistical analysis
excluded_columns = ["C_peptide", "IL6", "Leptin", "MCP1"]

# Calculate missing percentage for each variable
missing_by_group = []

for i, col in enumerate(df.columns):

    # Skip metadata columns
    if col in ["Participant_ID", "Treatment", "Measurement_Time"]:
        continue

    # Skip the four variables already excluded
    if col in excluded_columns:
        continue

    # Select the exact column by position
    values = df.iloc[:, i]

    # Calculate overall missing percentage
    overall_missing = values.isna().mean() * 100

    # Keep only variables with at least 30% missing values
    if overall_missing >= 30:

        row = {
            "Variable": col,
            "Overall_Missing_Percent": overall_missing
        }

        # Calculate missing percentage for each treatment × time group
        for treatment in df["Treatment"].unique():
            for time in sorted(df["Measurement_Time"].unique()):

                group = df[
                    (df["Treatment"] == treatment) &
                    (df["Measurement_Time"] == time)
                ]

                row[f"{treatment}_{time}"] = (
                    values.loc[group.index].isna().mean() * 100
                )

        missing_by_group.append(row)

# Create table
missing_by_group = pd.DataFrame(missing_by_group)

# Sort from highest to lowest overall missingness
missing_by_group = (
    missing_by_group
    .sort_values("Overall_Missing_Percent", ascending=False)
    .reset_index(drop=True)
)

missing_by_group

,Variable,Overall_Missing_Percent,Placebo_0,Placebo_3,SH120_0,SH120_3,SH60_0,SH60_3
0,IL1b,38.175676,0.0,0.0,12.0,22.0,100.0,100.0
1,IL1RA,35.810811,0.0,0.0,8.0,12.0,100.0,100.0
2,IL10,35.135135,0.0,0.0,6.0,10.0,100.0,100.0
3,Fetuin_A,33.108108,0.0,0.0,2.0,2.0,100.0,100.0
4,CRP,33.108108,0.0,0.0,2.0,2.0,100.0,100.0
5,Glucagon,33.108108,0.0,0.0,2.0,2.0,100.0,100.0
6,GLP1_total,33.108108,0.0,0.0,2.0,2.0,100.0,100.0
7,Adipsin,33.108108,0.0,0.0,2.0,2.0,100.0,100.0
8,TNFa,33.108108,0.0,0.0,2.0,2.0,100.0,100.0
9,IFNg,33.108108,0.0,0.0,2.0,2.0,100.0,100.0


### Variables with ≥30% Missing Values

The variables listed above have >30% overall missing values because no measurements are available for the SH60 treatment group. 
Measurements are available for the Placebo and SH120 groups, although some missing values are also present in SH120. Therefore, these variables will not be excluded entirely. Instead, SH60 will be excluded from analyses involving these variables, while the available Placebo and SH120 measurements will be retained.

### Data Types

Categorical Variables
Let's review the categorical variables to ensure that values are consistently coded and that no unexpected categories are present.

In [182]:
# Validate Treatment categories
print(df["Treatment"].value_counts(dropna=False))

Treatment
Placebo    100
SH120      100
SH60        96
Name: count, dtype: int64


In [183]:
# Validate categorical and identifier variables
for col in ["Treatment", "Gender", "Phototype", "Measurement_Time", "Participant_ID"]:
    print(f"\n{col}:")
    print(df[col].value_counts(dropna=False).to_string())


Treatment:
Treatment
Placebo    100
SH120      100
SH60        96

Gender:
Gender
Female    190
Male      106

Phototype:
Phototype
2    266
1     16
3     14

Measurement_Time:
Measurement_Time
0    148
3    148

Participant_ID:
Participant_ID
101    2
106    2
111    2
112    2
115    2
117    2
126    2
129    2
131    2
136    2
137    2
138    2
142    2
144    2
145    2
146    2
149    2
001    2
002    2
006    2
011    2
013    2
014    2
015    2
017    2
018    2
019    2
021    2
024    2
026    2
027    2
028    2
031    2
032    2
039    2
040    2
043    2
047    2
050    2
052    2
054    2
056    2
070    2
074    2
078    2
088    2
090    2
091    2
099    2
147    2
100    2
102    2
103    2
107    2
109    2
110    2
114    2
118    2
119    2
123    2
124    2
125    2
132    2
133    2
134    2
135    2
141    2
143    2
004    2
005    2
009    2
010    2
020    2
023    2
025    2
033    2
037    2
041    2
042    2
044    2
046    2
048    2
051    2
053    

In [184]:
# Check pandas data types of categorical and identifier variables
categorical_variables = [
    "Treatment",
    "Gender",
    "Phototype",
    "Measurement_Time",
    "Participant_ID"
]

df[categorical_variables].dtypes

Treatment              str
Gender              object
Phototype           object
Measurement_Time    object
Participant_ID      object
dtype: object

### Quantitative Variables

Let's review the data types of the quantitative variables to identify measurements that are not stored as numeric values where appropriate.

In [185]:
# Define quantitative variables from the current dataframe
quantitative_variables = [
    col for col in df.columns
    if col not in [
        "Treatment",
        "Gender",
        "Phototype",
        "Measurement_Time",
        "Participant_ID"
    ]
]

print("Total columns:", df.shape[1])
print("Quantitative variables:", len(quantitative_variables))

Total columns: 782
Quantitative variables: 777


In [186]:
# Check pandas data types of quantitative variables
df[quantitative_variables].dtypes

Age                  object
Alanin               object
Arginine             object
Asparagine           object
Citruline            object
                      ...  
Undecylenic_acid     object
Uric_acid            object
Uridine              object
Xanthine             object
Β_Muricholic_acid    object
Length: 801, dtype: object

In [187]:
print("Age:")
print(df["Age"].head(10).to_string(index=False))
print("\nAge dtype:", df["Age"].dtype)

print("\nAlanin:")
print(df["Alanin"].head(10).to_string(index=False))
print("\nAlanin dtype:", df["Alanin"].dtype)

Age:
57
57
55
34
55
35
42
25
34
45

Age dtype: object

Alanin:
 294.324357
        NaN
 420.792795
 123.192424
 746.432357
 471.293114
 594.824998
        NaN
 399.289288
 608.707767

Alanin dtype: object


In [ ]:
print("Age convertible to numeric:",
      pd.to_numeric(df["Age"], errors="coerce").notna().sum())

print("Alanin convertible to numeric:",
      pd.to_numeric(df["Alanin"], errors="coerce").notna().sum())

In [188]:
# Scan quantitative columns for non-numeric, non-missing values
non_numeric_values = []

for i, col in enumerate(df.columns):

    # Skip categorical and identifier variables
    if col in [
        "Treatment",
        "Gender",
        "Phototype",
        "Measurement_Time",
        "Participant_ID"
    ]:
        continue

    # Select the physical column by position
    values = df.iloc[:, i]

    # Identify non-missing values that cannot be converted to numeric
    converted = pd.to_numeric(values, errors="coerce")

    invalid = values.notna() & converted.isna()

    # Store every problematic value
    for row_index in df.index[invalid]:
        non_numeric_values.append({
            "Column_Position": i,
            "Variable": col,
            "Row": row_index,
            "Value": values.loc[row_index]
        })

non_numeric_values = pd.DataFrame(non_numeric_values)

print("Number of non-numeric, non-missing values:",
      len(non_numeric_values))

non_numeric_values

Number of non-numeric, non-missing values: 0


""


In [189]:
# Identify quantitative columns by their physical positions
quantitative_positions = [
    i for i, col in enumerate(df.columns)
    if col not in [
        "Treatment",
        "Gender",
        "Phototype",
        "Measurement_Time",
        "Participant_ID"
    ]
]

print("Quantitative columns:", len(quantitative_positions))

Quantitative columns: 777


In [190]:
# Display the data type of each quantitative column
for i in quantitative_positions:
    print(f"{i}: {df.columns[i]} -> {df.iloc[:, i].dtype}")

5: Age -> object
6: Alanin -> object
7: Arginine -> object
8: Asparagine -> object
9: Citruline -> object
10: Fenylalanin -> object
11: Glutamic_acid -> object
12: Glutamin -> object
13: Glycine -> object
14: Histidin -> object
15: Isoleucine -> object
16: Aspartic_Acid -> object
17: Leucine -> object
18: Lysin -> object
19: Methionine -> object
20: Ornithine -> object
21: Proline -> object
22: Pyrrolidone_carboxylic_acid -> object
23: Serine -> object
24: Threonine -> object
25: Tryptofan -> object
26: Tyrosine -> object
27: Urea -> object
28: Valin -> object
29: NMF -> object
30: Total_urocanic_acid -> object
31: Trans_Urocanic_acid -> object
32: Hydration_forehead -> object
33: Hydration_cheek -> object
34: Hydration_forearm -> object
35: TEWL_forehead -> object
36: TEWL_Cheek -> object
37: Sebum_forehead -> object
38: Sebum_cheek -> object
39: Wrinkle -> object
40: Red_Area_forehead -> object
41: Red_Area_cheek -> object
42: Pores_Area_forehead -> object
43: Pores_Area_cheek -> obj

In [191]:
# Convert quantitative columns to numeric by physical position
for i in quantitative_positions:
    df.iloc[:, i] = pd.to_numeric(
        df.iloc[:, i],
        errors="coerce"
    )

In [192]:
# Verify the data types after conversion
for i in quantitative_positions[:20]:
    print(f"{i}: {df.columns[i]} -> {df.iloc[:, i].dtype}")

5: Age -> object
6: Alanin -> object
7: Arginine -> object
8: Asparagine -> object
9: Citruline -> object
10: Fenylalanin -> object
11: Glutamic_acid -> object
12: Glutamin -> object
13: Glycine -> object
14: Histidin -> object
15: Isoleucine -> object
16: Aspartic_Acid -> object
17: Leucine -> object
18: Lysin -> object
19: Methionine -> object
20: Ornithine -> object
21: Proline -> object
22: Pyrrolidone_carboxylic_acid -> object
23: Serine -> object
24: Threonine -> object


In [193]:
test_age = pd.to_numeric(df.iloc[:, 5], errors="coerce")

print("Original dtype:", df.iloc[:, 5].dtype)
print("Converted dtype:", test_age.dtype)

Original dtype: object
Converted dtype: int64


In [194]:
# Convert all quantitative columns to numeric by position
for i in quantitative_positions:
    df[df.columns[i]] = pd.to_numeric(df.iloc[:, i], errors="coerce")

In [195]:
# Convert quantitative columns by physical position
for i in quantitative_positions:
    df.iloc[:, i] = pd.to_numeric(df.iloc[:, i], errors="coerce")

In [196]:
for i in quantitative_positions:
    converted = pd.to_numeric(df.iloc[:, i], errors="coerce")
    df.iloc[:, i] = converted.to_numpy()

In [197]:
print(df.iloc[:, 5].dtype)

int64


In [198]:
# Summarize data types of all quantitative columns
dtype_summary = {}

for i in quantitative_positions:
    dtype = str(df.iloc[:, i].dtype)
    dtype_summary[dtype] = dtype_summary.get(dtype, 0) + 1

print(dtype_summary)

{'int64': 1, 'float64': 776}


In [199]:
print(df[
    ["Treatment", "Gender", "Phototype",
     "Measurement_Time", "Participant_ID"]
].dtypes)

Treatment              str
Gender              object
Phototype           object
Measurement_Time    object
Participant_ID      object
dtype: object


The data types of the dataset were reviewed and standardized according to the intended role of each variable. The five categorical and identifier variables were retained in their original non-numeric formats, while all 777 quantitative variables were successfully converted to numeric types. One quantitative variable (`Age`) is stored as `int64`, while the remaining 776 quantitative variables are stored as `float64`. No non-missing values were found that could not be interpreted as numeric.

### Data Consistency

Let's review the dataset for unexpected or inconsistent values and identify potential extreme observations. Extreme values will be investigated as possible data-entry errors, measurement errors, or valid observations. No clinical measurements will be changed without a documented methodological reason.

In [200]:
# Identify potential outliers using the IQR rule
potential_outliers = []

for i in quantitative_positions:
    values = df.iloc[:, i]

    q1 = values.quantile(0.25)
    q3 = values.quantile(0.75)
    iqr = q3 - q1

    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr

    outlier_count = ((values < lower_bound) | (values > upper_bound)).sum()

    potential_outliers.append({
        "Column_Position": i,
        "Variable": df.columns[i],
        "Lower_Bound": lower_bound,
        "Upper_Bound": upper_bound,
        "Potential_Outliers": outlier_count
    })

potential_outliers = pd.DataFrame(potential_outliers)

potential_outliers

,Column_Position,Variable,Lower_Bound,Upper_Bound,Potential_Outliers
0,5,Age,1.312500e+01,7.412500e+01,0
1,6,Alanin,2.221823e+01,1.210760e+03,5
2,7,Arginine,-1.927401e+02,9.885313e+02,1
3,8,Asparagine,5.957503e+01,2.209687e+02,4
4,9,Citruline,-8.767171e+01,1.366651e+03,6
...,...,...,...,...,...
772,777,Undecylenic_acid,-8.640693e+05,5.438092e+06,8
773,778,Uric_acid,8.315064e+08,1.947367e+09,6
774,779,Uridine,3.636599e+06,1.313769e+07,16
775,780,Xanthine,2.298678e+06,1.029675e+07,21


In [201]:
# Show variables with potential outliers
potential_outliers[
    potential_outliers["Potential_Outliers"] > 0
].sort_values(
    "Potential_Outliers",
    ascending=False
)

,Column_Position,Variable,Lower_Bound,Upper_Bound,Potential_Outliers
679,684,M_4_Chlorophenoxyacetic_acid,1.766321e+04,4.466255e+04,53
686,691,Acetaminophen_glucuronide,1.021835e+03,6.675256e+04,50
767,772,Ser_Leu_Tyr,2.824477e+04,4.332755e+04,50
771,776,UKOUEEGEHXWHHO_MOFOKWOHSA_N,7.118825e+05,1.687976e+06,45
725,730,Cholic_Acid,-2.770669e+06,5.753393e+06,42
...,...,...,...,...,...
64,69,Elasticity_R5_cheek,-1.567583e-01,7.999750e-01,1
46,51,MEXA_Erythema_cheek,2.035529e+02,6.629452e+02,1
32,37,Sebum_forehead,-8.500000e+00,2.115000e+02,1
754,759,N_Formylmethionine,8.225520e+05,5.669834e+06,1


In [202]:
# Check for actual negative values in quantitative variables
negative_values = []

for i in quantitative_positions:
    values = df.iloc[:, i]
    count = (values < 0).sum()

    if count > 0:
        negative_values.append({
            "Variable": df.columns[i],
            "Negative_Count": count
        })

negative_values = pd.DataFrame(negative_values)

print("Variables containing negative values:")
print(negative_values)

Variables containing negative values:
                 Variable  Negative_Count
0               ITA_cheek               3
1     Elasticity_F0_cheek               1
2  Elasticity_F1_forehead               1
3                    TNFa               2
4                    IL1b              32
5                    IL10               4
6                   IL17A              14
7                     IL6              15


In [203]:
# Show the actual negative values
for variable in negative_values["Variable"]:
    values = df.loc[df[variable] < 0, ["Participant_ID", "Treatment", "Measurement_Time", variable]]
    
    print(f"\n{variable}:")
    print(values.to_string(index=False))


ITA_cheek:
Participant_ID Treatment Measurement_Time  ITA_cheek
           048     SH120                0  -6.615385
           121      SH60                0  -1.000000
           148      SH60                0  -5.800000

Elasticity_F0_cheek:
Participant_ID Treatment Measurement_Time  Elasticity_F0_cheek
           014   Placebo                3            -0.000167

Elasticity_F1_forehead:
Participant_ID Treatment Measurement_Time  Elasticity_F1_forehead
           016      SH60                3                 -0.0002

TNFa:
Participant_ID Treatment Measurement_Time      TNFa
           085     SH120                0 -2.715600
           085     SH120                3 -1.709467

IL1b:
Participant_ID Treatment Measurement_Time      IL1b
           115   Placebo                0 -1.401039
           126   Placebo                0 -0.631660
           131   Placebo                0 -1.016349
           136   Placebo                0 -0.246971
           144   Placebo                0

Negative cytokine values may reflect assay-related effects such as low-level measurements or standard-curve calibration. Previous studies have shown that background subtraction in multiplex immunoassays can produce negative assay values, particularly when measured signals are close to or below the background level. These values were therefore retained rather than treated automatically as data-entry errors.

As regards `ITA_cheek`, negative values are expected for this measurement, while the negative values observed for `Elasticity_F0_cheek` and `Elasticity_F1_forehead` are nearly zero and were therefore also retained.

Conclusion: all negative values will be retained for statistical analysis and will not be removed or modified solely because they are negative.

In [204]:
# Identify potential outliers by treatment group
outliers_by_treatment = []

for i in quantitative_positions:
    values = df.iloc[:, i]

    q1 = values.quantile(0.25)
    q3 = values.quantile(0.75)
    iqr = q3 - q1

    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr

    outlier_mask = (values < lower_bound) | (values > upper_bound)

    if outlier_mask.sum() > 0:
        for treatment in df["Treatment"].unique():
            count = (
                outlier_mask &
                (df["Treatment"] == treatment)
            ).sum()

            if count > 0:
                outliers_by_treatment.append({
                    "Variable": df.columns[i],
                    "Treatment": treatment,
                    "Potential_Outliers": count
                })

outliers_by_treatment = pd.DataFrame(outliers_by_treatment)

outliers_by_treatment.sort_values(
    "Potential_Outliers",
    ascending=False
)

,Variable,Treatment,Potential_Outliers
1832,M_4_Chlorophenoxyacetic_acid,SH60,22
1692,Fetuin_A,Placebo,21
1851,Acetaminophen_glucuronide,Placebo,19
1963,Cholic_Acid,Placebo,18
1761,M_2_5_Dihydroxybenzenesulfonic_Acid,Placebo,18
...,...,...,...
570,LPE_18_0,SH60,1
593,LPI_18_0,SH60,1
41,Pyrrolidone_carboxylic_acid,SH120,1
35,Ornithine,Placebo,1


In [206]:
# Calculate the percentage of potential outliers for each variable
outlier_frequency = []

for i in quantitative_positions:
    values = df.iloc[:, i]

    q1 = values.quantile(0.25)
    q3 = values.quantile(0.75)
    iqr = q3 - q1

    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr

    outlier_mask = (
        (values < lower_bound) |
        (values > upper_bound)
    )

    non_missing = values.notna().sum()
    outlier_count = outlier_mask.sum()

    outlier_percent = (
        outlier_count / non_missing * 100
    )

    outlier_frequency.append({
        "Column_Position": i,
        "Variable": df.columns[i],
        "Non_Missing": non_missing,
        "Potential_Outliers": outlier_count,
        "Outlier_Percent": outlier_percent
    })

outlier_frequency = pd.DataFrame(outlier_frequency)

In [207]:
# Divide variables according to the proportion of potential outliers
outlier_frequency["Outlier_Group"] = pd.cut(
    outlier_frequency["Outlier_Percent"],
    bins=[-np.inf, 20, 40, np.inf],
    labels=["<20%", "20%–40%", ">40%"],
    right=False
)

outlier_frequency["Outlier_Group"].value_counts()

Outlier_Group
<20%       777
20%–40%      0
>40%         0
Name: count, dtype: int64

In [208]:
outlier_frequency["Potential_Outliers"].value_counts().sort_index()

Potential_Outliers
0      5
1     16
2     22
3     28
4     44
5     42
6     38
7     53
8     46
9     42
10    53
11    41
12    41
13    37
14    36
15    30
16    24
17    20
18    19
19    14
20    16
21    11
22    11
23    14
24    12
25     5
26    11
27    11
28     4
29     4
30     5
31     2
32     1
33     3
34     1
35     4
37     2
38     1
40     1
41     2
42     1
45     1
50     2
53     1
Name: count, dtype: int64

In [209]:
# Group variables according to the number of potential outliers
outlier_frequency["Outlier_Group"] = pd.cut(
    outlier_frequency["Potential_Outliers"],
    bins=[-1, 0, 5, 10, 20, np.inf],
    labels=[
        "0",
        "1–5",
        "6–10",
        "11–20",
        ">20"
    ]
)

outlier_frequency["Outlier_Group"].value_counts().sort_index()

Outlier_Group
0          5
1–5      152
6–10     232
11–20    278
>20      110
Name: count, dtype: int64

In [210]:
# Review variables with more than 20 potential outliers
high_outlier_variables = outlier_frequency[
    outlier_frequency["Outlier_Group"] == ">20"
].copy()

high_outlier_variables

,Column_Position,Variable,Non_Missing,Potential_Outliers,Outlier_Percent,Outlier_Group
7,12,Glutamin,288,27,9.375000,>20
25,30,Total_urocanic_acid,288,22,7.638889,>20
35,40,Red_Area_forehead,295,24,8.135593,>20
36,41,Red_Area_cheek,294,35,11.904762,>20
38,43,Pores_Area_cheek,295,22,7.457627,>20
...,...,...,...,...,...,...
766,771,Ser_Glu,292,26,8.904110,>20
767,772,Ser_Leu_Tyr,292,50,17.123288,>20
768,773,Succinic_acid,292,22,7.534247,>20
771,776,UKOUEEGEHXWHHO_MOFOKWOHSA_N,292,45,15.410959,>20


In [211]:
# Review variables with more than 20 potential outliers
high_outlier_variables = outlier_frequency[
    outlier_frequency["Outlier_Group"] == ">20"
].copy()

# Add outlier counts for each treatment group
for treatment in ["Placebo", "SH60", "SH120"]:
    counts = []

    for _, row in high_outlier_variables.iterrows():
        i = row["Column_Position"]
        values = df.iloc[:, i]

        q1 = values.quantile(0.25)
        q3 = values.quantile(0.75)
        iqr = q3 - q1

        lower_bound = q1 - 1.5 * iqr
        upper_bound = q3 + 1.5 * iqr

        outlier_mask = (
            (values < lower_bound) |
            (values > upper_bound)
        )

        count = (
            outlier_mask &
            (df["Treatment"] == treatment)
        ).sum()

        counts.append(count)

    high_outlier_variables[treatment] = counts

# Display the complete table
print(high_outlier_variables.to_string(index=False))

 Column_Position                                                                                                  Variable  Non_Missing  Potential_Outliers  Outlier_Percent Outlier_Group  Placebo  SH60  SH120
              12                                                                                                  Glutamin          288                  27         9.375000           >20        8    10      9
              30                                                                                       Total_urocanic_acid          288                  22         7.638889           >20        6     7      9
              40                                                                                         Red_Area_forehead          295                  24         8.135593           >20        6     8     10
              41                                                                                            Red_Area_cheek          294                  35        1

The variables with more than 20 potential outliers were reviewed according to treatment group. For most variables, observations were distributed across Placebo, SH60, and SH120 rather than restricted to a single treatment group. This pattern does not provide evidence that the observations are data-entry errors. They were therefore considered potential extreme observations and retained for statistical analysis.

In [212]:
# Review variables with 11–20 potential outliers
medium_outlier_variables = outlier_frequency[
    outlier_frequency["Outlier_Group"] == "11–20"
].copy()

# Add potential outlier counts by treatment group
for treatment in ["Placebo", "SH60", "SH120"]:
    counts = []

    for _, row in medium_outlier_variables.iterrows():
        i = row["Column_Position"]
        values = df.iloc[:, i]

        q1 = values.quantile(0.25)
        q3 = values.quantile(0.75)
        iqr = q3 - q1

        lower_bound = q1 - 1.5 * iqr
        upper_bound = q3 + 1.5 * iqr

        outlier_mask = (
            (values < lower_bound) |
            (values > upper_bound)
        )

        count = (
            outlier_mask &
            (df["Treatment"] == treatment)
        ).sum()

        counts.append(count)

    medium_outlier_variables[treatment] = counts

print(medium_outlier_variables.to_string(index=False))

 Column_Position                           Variable  Non_Missing  Potential_Outliers  Outlier_Percent Outlier_Group  Placebo  SH60  SH120
              11                      Glutamic_acid          288                  11         3.819444         11–20        1     7      3
              17                            Leucine          288                  14         4.861111         11–20        3     8      3
              18                              Lysin          288                  12         4.166667         11–20        2     5      5
              27                               Urea          288                  13         4.513889         11–20        3     8      2
              28                              Valin          288                  12         4.166667         11–20        3     6      3
              36                         TEWL_Cheek          295                  15         5.084746         11–20        4     5      6
              48              MEXA

For this group. Although some variables showed differences in the number of outlier observations between treatment groups, this pattern alone does not provide evidence of data entry or measurement errors. 

In [213]:
outliers_6_10 = outlier_frequency[
    outlier_frequency["Outlier_Group"] == "6–10"
].copy()

outliers_6_10.shape

(232, 6)

In [214]:
outliers_6_10 = outlier_frequency[
    outlier_frequency["Outlier_Group"] == "6–10"
].copy()

for treatment in df["Treatment"].unique():
    outliers_6_10[treatment] = 0

for i in outliers_6_10.index:
    position = outliers_6_10.loc[i, "Column_Position"]
    values = df.iloc[:, position]

    q1 = values.quantile(0.25)
    q3 = values.quantile(0.75)
    iqr = q3 - q1

    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr

    outlier_mask = (values < lower_bound) | (values > upper_bound)

    for treatment in df["Treatment"].unique():
        outliers_6_10.loc[i, treatment] = (
            outlier_mask & (df["Treatment"] == treatment)
        ).sum()

outliers_6_10 = outliers_6_10[
    [
        "Variable",
        "Non_Missing",
        "Potential_Outliers",
        "Outlier_Percent",
        "Placebo",
        "SH60",
        "SH120"
    ]
]

print(outliers_6_10.to_string(index=False))

                                 Variable  Non_Missing  Potential_Outliers  Outlier_Percent  Placebo  SH60  SH120
                                Citruline          288                   6         2.083333        0     2      4
                              Fenylalanin          288                  10         3.472222        0     6      4
                               Isoleucine          288                   7         2.430556        1     4      2
                            Aspartic_Acid          288                   7         2.430556        2     4      1
                               Methionine          288                   7         2.430556        1     4      2
                                Ornithine          288                  10         3.472222        1     5      4
                                  Proline          288                  10         3.472222        2     5      3
              Pyrrolidone_carboxylic_acid          288                   6         2.083

Some variables showed differences in the number of observations between treatment groups, but this pattern again does not provide evidence of data entry or measurement errors. 

## Conclusion

The clinical trial dataset was successfully cleaned and validated in preparation for statistical analysis. The process included standardizing the Excel structure and variable names, removing redundant information, checking data types, assessing missing values and duplicates, and validating participant and measurement-time records. Potential outliers were also investigated without automatically removing observations solely because they were extreme.

Overall, the cleaning and validation process provided a more consistent and structured dataset suitable for the next stage of the project. The resulting cleaned dataset will be used for statistical analysis of the clinical trial data.